### PR TITLE
trainer from dataset fetch targets

### PR DESCRIPTION
**Pull request by seiriosPlus**
添加executor.FetchHandler类，用户 trainer_from_dataset/infer_from_dataset的输出优化。

使用方法：
用户自定义Class：
```
        class FH(fluid.executor.FetchHandler):
            def handler(self, fetch_target_vars):
                g_auc = fetch_target_vars[0]
                print("g_auc: {}, time: {}".format(np.mean(g_auc), time.ctime()))
```

在trainer_from_dataset中使用：
```
            exe.train_from_dataset(
                program=fleet.main_program,
                dataset=dataset,
                fetch_handler=FH(fetch_target_names=[auc_var.name], period_secs=2),
                debug=False)
```

使用效果：
image

### CODE REVIEW

File: python/paddle/fluid/executor.py [hunk 1]
```diff
@@ -814,6 +836,67 @@ def _prepare_trainer(self,
...
+            thread = self._adjust_pipeline_resource(program._pipeline_opt,
+                                                    dataset, thread)
+
+        dataset._prepare_to_run()
+
+        if fetch_handler is not None:
+            fetch_instance = fetch_handler
+        elif fetch_handler is None and fetch_list is not None:
+
+            class FH(FetchHandler):
```
Comment (guru4elephant) [inline python/paddle/fluid/executor.py:967]: is this handler only used for variable print? what if I want to do some calculation based on fetched variables?

Comment (seiriosPlus) [reply python/paddle/fluid/executor.py:967]: you can do anything you want based on fetched variables.

**Review by frankwhzhang — APPROVED** [review]
Doc LGTM

**Review by kuke — APPROVED** [review]
Approve for API change

**Pull request merged**